lnbits/extensions/boltz/boltz.py

```diff
@@ -0,0 +1,432 @@
+logger.debug(f"BOLTZ_URL: {BOLTZ_URL}")
+logger.debug(f"Bitcoin Network: {net['name']}")
+
+
+async def create_swap(data: CreateSubmarineSwap) -> SubmarineSwap:
+
+    if not check_boltz_limits(data.amount):
+        msg = f"Boltz - swap not in boltz limits"
+        logger.warning(msg)
+        raise Exception(msg)
+
+    swap_id = urlsafe_short_hash()
+    try:
+        payment_hash, payment_request = await create_invoice(
+            wallet_id=data.wallet,
+            amount=data.amount,
+            memo=f"swap of {data.amount} sats on boltz.exchange",
+            extra={"tag": "boltz", "swap_id": swap_id},
+        )
+    except Exception as exc:
+        msg = f"Boltz - create_invoice failed {str(exc)}"
+        logger.error(msg)
+        raise
+
+    refund_privkey = ec.PrivateKey(os.urandom(32), True, net)
+    refund_pubkey_hex = hexlify(refund_privkey.sec()).decode("UTF-8")
+
+    res = req_wrap(
+        "post",
+        f"{BOLTZ_URL}/createswap",
+        json={
+            "type": "submarine",
+            "pairId": "BTC/BTC",
+            "orderSide": "sell",
+            "refundPublicKey": refund_pubkey_hex,
+            "invoice": payment_request,
+            "referralId": "lnbits",
+        },
+        headers={"Content-Type": "application/json"},
+    )
+    res = res.json()
+    logger.info(
+        f"Boltz - created normal swap, boltz_id: {res['id']}. wallet: {data.wallet}"
+    )
+    return SubmarineSwap(
+        id=swap_id,
+        time=get_timestamp(),
+        wallet=data.wallet,
+        amount=data.amount,
+        payment_hash=payment_hash,
+        refund_privkey=refund_privkey.wif(net),
+        refund_address=data.refund_address,
+        boltz_id=res["id"],
+        status="pending",
+        address=res["address"],
+        expected_amount=res["expectedAmount"],
+        timeout_block_height=res["timeoutBlockHeight"],
+        bip21=res["bip21"],
+        redeem_script=res["redeemScript"],
+    )
+
+
+"""
+explanation taken from electrum
+send on Lightning, receive on-chain
+- User generates preimage, RHASH. Sends RHASH to server.
+- Server creates an LN invoice for RHASH.
+- User pays LN invoice - except server needs to hold the HTLC as preimage is unknown.
+- Server creates on-chain output locked to RHASH.
+- User spends on-chain output, revealing preimage.
+- Server fulfills HTLC using preimage.
+Note: expected_onchain_amount_sat is BEFORE deducting the on-chain claim tx fee.
+"""
+
+
+async def create_reverse_swap(
+    data: CreateReverseSubmarineSwap,
+) -> [ReverseSubmarineSwap, asyncio.Task]:
+
+    if not check_boltz_limits(data.amount):
+        msg = f"Boltz - reverse swap not in boltz limits"
+        logger.warning(msg)
+        raise Exception(msg)
+
+    swap_id = urlsafe_short_hash()
+
+    if not await check_balance(data):
+        logger.error(
+            f"Boltz - reverse swap, insufficient balance. create_reverse_swap should never be called before balance checking."
```

Comment:
```suggestion
            f"Boltz - reverse swap, insufficient balance."
```

This error is user-facing, it doesn't really make sense to tell them "create_reverse_swap should never be called before balance checking." – why not put the balance check into create_reverse_swap?

Reply:
my idea behind this was, although it maybe not a good idea. the user has only the api exposed to him, where i do the balance check. but if so another ext or anything else uses it (like tests) i double check it and give this error.

i think i will change that, i was an afterthought when i was writing tests, if i use it in both cases i might aswell only check there.

Reply:
31546088